Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigation.swift

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2024 Apple Inc. All rights reserved.
+ //
+ // Redistribution and use in source and binary forms, with or without
+ // modification, are permitted provided that the following conditions
+ // are met:
+ // 1. Redistributions of source code must retain the above copyright
+ //    notice, this list of conditions and the following disclaimer.
+ // 2. Redistributions in binary form must reproduce the above copyright
+ //    notice, this list of conditions and the following disclaimer in the
+ //    documentation and/or other materials provided with the distribution.
+ //
+ // THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ // AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ // THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ // PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ // BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ // CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ // SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ // INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ // CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ // ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ // THE POSSIBILITY OF SUCH DAMAGE.
+
+#if canImport(Testing) && compiler(>=6.0)
+
+import Testing
+import WebKit
+@_spi(Internal) import WebKit
+
+// MARK: Helper extension functions
+
+fileprivate extension Array {
+    init<Failure>(_ sequence: some AsyncSequence<Self.Element, Failure>) async throws(Failure) where Failure: Error {
+        self.init()
+
+        for try await element in sequence {
+            self.append(element)
+        }
+    }
+}
+
+fileprivate extension URL {
+    static var aboutBlank: URL {
+        URL(string: "about:blank")!
+    }
+}
+
+// MARK: Tests
+
+@MainActor
+struct WebPageNavigationTests {
+    @Test
+    func basicNavigation() async throws {
+        let page = WebPage()
+        
+        let html = """
+        <html>
+        <div>Hello</div>
+        </html>
+        """
+        
+        let expectedEvents: [WebPage.NavigationEvent] = [.startedProvisionalNavigation, .committed, .finished]
+        
+        let navigation = page.navigate(to: html, baseURL: .aboutBlank)
+        let actualEvents = try await Array(navigation)
+        
+        #expect(actualEvents == expectedEvents)
+    }
+    
+    @Test
+    func failedProvisionNavigationThrowsError() async throws {
+        let page = WebPage()
+        
+        let request = URLRequest(url: URL(string: "about:foo")!)
+        
+        await #expect(throws: WebPage.NavigationError.self) {
+            for try await _ in page.navigate(to: request) {
+            }
```

Comment:
Hm... this is an interesting API shape!

I'm curious what happens in the cases where:

- The web content process is terminated during navigation (does that invoke `-didFailNavigation`? or should the Swift equivalent of `-webViewWebContentProcessDidTerminate:` be treated as a "terminal" state, similar to `-didFinishNavigation`)
- How would a SwiftUI WebKit client respond correctly to navigation that's either programmatically triggered by the page or triggered by the user? (e.g. the client wants to be informed when a load is committed so that it can show an indeterminate progress indicator in the UI)